lib/auth/auth.go

```diff
@@ -5384,6 +5388,98 @@ func (a *Server) CompareAndSwapHeadlessAuthentication(ctx context.Context, old,
 	return headlessAuthn, trace.Wrap(err)
 }
 
+// GetResourceUsage returns the usage data for resources which are limited on usage-based billing plans.
+func (a *Server) GetResourceUsage(ctx context.Context, req *proto.GetResourceUsageRequest) (*proto.GetResourceUsageResponse, error) {
+	features := modules.GetModules().Features()
+	if !features.IsUsageBasedBilling {
+		return &proto.GetResourceUsageResponse{}, nil
+	}
+	accessRequestUsage, err := a.getAccessRequestMonthlyUsage(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &proto.GetResourceUsageResponse{
+		AccessRequests: accessRequestUsage,
+	}, nil
+}
+
+// getAccessRequestMonthlyUsage returns the number of access requests that have been used this month.
+// Only requests that were both created and since the start of the current month are included in this number.
+func (a *Server) getAccessRequestMonthlyUsage(ctx context.Context) (*proto.AccessRequestUsage, error) {
+	features := modules.GetModules().Features()
+	monthlyLimit := features.AccessRequests.MonthlyRequestLimit
+
+	now := a.clock.Now().UTC()
+	monthStart := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
+
+	created := make(map[string]struct{})
+	reviewed := make(map[string]struct{})
+
+	var results []apievents.AuditEvent
+	var startKey string
+	var err error
+	for {
+		results, startKey, err = a.SearchEvents(ctx, events.SearchEventsRequest{
```

Comment:
Do we know how expensive that call is?

Reply:
It helps that we have an index on event creation time. I did rudimentary manual tests on a busy-ish production server and the query consistently took under 300 ms. As this is only applicable on the Team plan, performance shouldn't be a big issue.